File: test_functionality.py
# import project module ala
import ala
import sirf.STIR as PET
from matplotlib import pyplot as plt
import numpy
from os.path import join
# read PET data examples
sirf_examples = PET.examples_data_path('PET')

#%% data reading/creation
# test creating a STIRImageData object with given dimensions etc
c= ala.STIRImageData()
c.initialise((40,5,6),(1,1,1));
print(c.get_geom_info_sptr().get_info())

path1 = join(sirf_examples,'test_image_PM_QP_6.hv')
c= ala.STIRImageData(path1) 
ctest= PET.ImageData(path1) 
print(c.get_geom_info_sptr().get_info())

path2 = join(sirf_examples,'Utahscat600k_ca_seg4.hs')
b= ala.PETAcquisitionDataInFile(path2)
btest= PET.AcquisitionData(path2)
#%% perform a simulation of a PET acquisition
#im=ala.STIRImageData(path1) 

# save figure
path3 = join(sirf_examples,'brain','emission.hv')
im= ala.STIRImageData(path3)
plt.imshow(im.as_array()[5,:,:])
filename = 'simu_graph.png'
plt.savefig(filename)
plt.close()

# construct a "template" for the acquisition data
sino_template= ala.PETAcquisitionDataInMemory("PRT-1",1,1)
#print(sino_template.get_info())

# construct an acquisition model and use it to simulate the data
acq_model = ala.PETAcquisitionModelUsingRayTracingMatrix();
acq_model.set_up(sino_template, im)
simulated_data = acq_model.forward(im)
#%% reconstruct with FBP2D
recon = ala.xSTIR_FBP2DReconstruction()
recon.set_input(simulated_data)
recon.set_up(im)
recon.process()   
image = recon.get_output()
# save figure
plt.imshow(image.as_array()[5,:,:])
filename = 'recon_graph.png'
plt.savefig(filename)
plt.close()

#%% plot some profiles
plt.figure()
plt.plot(im.as_array()[5,75,:])
plt.plot(image.as_array()[5,75,:])
plt.legend(["input", "reconstruction"])
plt.grid(True)
filename = 'profile_graph.png'
plt.savefig(filename)
plt.close()
#%%% more validation

# test ImageData wrapping
#c.fill(0)
print(c.dot(c))
print(ctest.dot(ctest))
print(c.norm())
print(ctest.norm())
print(c.is_complex())
print(ctest.is_complex())
# test AcquisitionData wrapping
print(b.norm())
print(btest.norm())
print(b.is_complex())
print(btest.is_complex())
# test as_array
print(c.as_array().shape)
print(ctest.as_array().shape)
print(ctest.as_array()[5,23,:])
plt.imshow(c.as_array()[5,:,:])
plt.grid(True)
filename = 'test_graph1.png'
plt.savefig(filename)
plt.close()
plt.imshow(ctest.as_array()[5,:,:])
plt.grid(True)
filename = 'test_graph2.png'
plt.savefig(filename)
plt.close()
# test from_array
print(c.from_array(c.as_array()*0))
print(c.dot(c))

File: test_functionality.m
%%
% addpath for the interface files
addpath '/home/sirfuser/devel/Rocci-SIRFPROJECT/SIRFPROJECT'
addpath '/home/sirfuser/devel/install/matlab/'
% launch PET
set_up_PET([]);
% get path for PET examples
sirf_examples_path = sirf.Utilities.examples_data_path('PET');
%%
% declare null STIRImageData and test initialise()
stir_data_null = ala.STIRImageData();
stir_data_null.initialise(40,5,6,1,1,1);
disp(stir_data_null.get_geom_info_sptr().get_info());
%%
% create STIRImageData from PET examples
stir_data_1 = ala.STIRImageData([sirf_examples_path,'/test_image_PM_QP_6.hv']);
disp(stir_data_1.get_geom_info_sptr().get_info());
stir_data_test = sirf.STIR.ImageData([sirf_examples_path,'/test_image_PM_QP_6.hv']);

%%
% create PETAcquisitionData from PET examples
acq_data = ala.PETAcquisitionDataInFile([sirf_examples_path,'/Utahscat600k_ca_seg4.hs']);
acq_data_test = sirf.STIR.AcquisitionData([sirf_examples_path,'/Utahscat600k_ca_seg4.hs']);

%%
% use as_array() to convert STIRImageData into an array
im = ala.STIRImageData([sirf_examples_path,'/brain/emission.hv']);
im_array = im.as_array();
im_array_one_pic = im_array(5,:,:);
%im_array_one_pic = reshape(im_array_one_pic,211,211);
%imshow(im_array_one_pic)
%% construct a "template" for the acquisition data
sino_template = ala.PETAcquisitionDataInMemory('PRT-1',1,1);


% construct an acquisition model and use it to simulate the data
acq_model = ala.PETAcquisitionModelUsingRayTracingMatrix();
acq_model.set_up(sino_template, im);
simulated_data = acq_model.forward(im);

%% reconstruct with FBP2D
recon = ala.FBP2DReconstructor();
recon.set_input(simulated_data);
recon.set_up(im);
recon.process();

%% get the output image of reconstructor
image_data = recon.get_output();
image_data.write('recon_test_image.hv');

%% test the SWIG-wrapped interfaces by comparing the result with manually wrapped functions
fprintf('the result of stir_data_1.dot(stir_data_1) is %g.\n', stir_data_1.dot(stir_data_1))
fprintf('the result of stir_data_test.dot(stir_data_test) is %g.\n', stir_data_test.dot(stir_data_test))
fprintf('the result of stir_data_1.norm() is %g.\n', stir_data_1.norm());
fprintf('the result of stir_data_test.norm() is %g.\n', stir_data_test.norm());
fprintf('the result of stir_data_1.is_complex() is %d (0 stands for false, 1 stands for true).\n', stir_data_1.is_complex());
%fprintf('the result of stir_data_test.is_complex() is %d (0 stands for false, 1 stands for true).\n', stir_data_test.is_complex());
fprintf('the result of acq_data.norm() is %g.\n', acq_data.norm());
fprintf('the result of acq_data_test.norm() is %g.\n', acq_data_test.norm());
fprintf('the reuslt of acq_data.is_complex() is %d (0 stands for false, 1 stands for true).\n', acq_data.is_complex());
%fprintf('the reuslt of acq_data_test.is_complex() is %d (0 stands for false, 1 stands for true).\n', acq_data_test.is_complex());

%% test as_array()
disp('the dimensions of stir_data_1 in each dimension is:');
disp(size(stir_data_1.as_array()));
disp('the dimensions of stir_data_test in each dimension is:');
disp(size(stir_data_test.as_array()));